### PR TITLE
4.x - Added ContextInterface::getPrimaryKey() to use instead of primaryKey()

### PR DESCRIPTION
**Pull request by othercorey**
Shouldn't we align this name with other getters()?

### CODE REVIEW

**Comment by ADmad** [comment]
I am not thrilled about unnecessary public API changes after 2 RCs. 

If we do want to want to add `getPrimaryKey()` then `primaryKey()` should be removed from the interface and kept only in the implementing classes with deprecated tag.

Same for #14017.

**Comment by ADmad** [comment]
I am not thrilled about unnecessary public API changes after 2 RCs. 

If we do want to want to add `getPrimaryKey()` then `primaryKey()` should be removed from the interface and kept only in the implementing classes with deprecated tag.

Same for #14017.

**Comment by othercorey** [comment]
> If we do want to want to add `getPrimaryKey()` then `primaryKey()` should be removed from the interface and kept only in the implementing classes with deprecated tag.

Agreed.  I'm only putting these up for awareness.  I defer to the team about what's correct.  I'll submit the requested changes.

**Comment by dereuromark** [comment]
My main problem with primaryKey() is its unusefulness by default for most cases, since you get string|array.
I would much rather have a getFirstPrimaryKey() or getPrimaryKeyScalarOrFail() kind of wrappers added instead that you can use directly in code (or fail early) instead of having to array_shift() or other operations on the result until you can trust the "primary" use case (99%) of dealing with a single primary key. Most of my apps and plugins always need a lot of overhead logic to deal with this.

Just adding get prefix here for the sake of naming doesn't solve too much here for most use cases.

**Comment by othercorey** [comment]
> Just adding get prefix here for the sake of naming doesn't solve too much here for most use cases.

Not dismissing your other points, but this is the reason I brought this up.  Reading this code, I thought what is different about `getPrimaryKey()` and `primaryKey()`?

This adds consistency which makes code more reliable.

```
class Table
{
    public function getPrimaryKey()
    {
        if ($this->_primaryKey === null) {
            $key = (array)$this->getSchema()->primaryKey();
            if (count($key) === 1) {
                $key = $key[0];
            }
            $this->_primaryKey = $key;
        }
```

**Comment by dereuromark** [comment]
Sry, so it wouldnt be a complete alias, but the said scalar return?
I didnt check the code exactly, so I didnt see this.

PS: https://github.com/cakephp/cakephp/blob/4.x/src/Http/Session/DatabaseSession.php#L155-L156
We are still lying everywhere here - quiet badly IMO.
It would be cleaner to have the correct way and otherwise fail with meaningful exception early.

File: src/View/Form/ArrayContext.php [hunk 1]
```diff
@@ -121,12 +122,34 @@ public function primaryKey(): array
         return [];
     }
 
+    /**
+     * Get the fields used in the context as a primary key.
+     *
+     * @return string[]
+     */
+    public function getPrimaryKey(): array
```
Comment (markstory) [inline src/View/Form/ArrayContext.php]: This looks like the same implementation as `primaryKey()` has. Why duplicate the code?

Comment (othercorey) [reply src/View/Form/ArrayContext.php]: Just an accident.  I left it in fixing a warning.

**Pull request merged**